src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseAsSpanInsteadOfRangeIndexer.cs

```diff
@@ -0,0 +1,211 @@
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+            var stringType = compilation.GetSpecialType(SpecialType.System_String);
+            var span = compilation.GetOrCreateTypeByMetadataName("System.Span`1");
+            var readOnlySpan = compilation.GetOrCreateTypeByMetadataName("System.ReadOnlySpan`1");
+            var memory = compilation.GetOrCreateTypeByMetadataName("System.Memory`1");
+            var readOnlyMemory = compilation.GetOrCreateTypeByMetadataName("System.ReadOnlyMemory`1");
+            var range = compilation.GetOrCreateTypeByMetadataName("System.Range");
+
+            if (stringType == null || span == null || readOnlySpan == null || memory == null || readOnlyMemory == null || range == null)
+            {
+                return;
+            }
+
+            var spanTypes = ImmutableArray.Create(span, readOnlySpan);
+            var readOnlyTypes = ImmutableArray.Create(readOnlySpan, readOnlyMemory);
+            var targetTypes = ImmutableArray.Create(span, readOnlySpan, memory, readOnlyMemory);
+
+            context.RegisterOperationAction(
+                operationContext =>
+                {
+                    IOperation indexerArgument;
+                    ITypeSymbol containingType;
+
+                    if (operationContext.Operation is IPropertyReferenceOperation propertyReference)
+                    {
+                        if (!propertyReference.Property.IsIndexer || propertyReference.Arguments.Length != 1)
+                        {
+                            return;
+                        }
+
+                        indexerArgument = propertyReference.Arguments[0].Value;
+                        containingType = propertyReference.Property.ContainingType;
+                    }
+                    else if (operationContext.Operation is IArrayElementReferenceOperation elementReference)
+                    {
+                        if (elementReference.Indices.Length != 1)
+                        {
+                            return;
+                        }
+
+                        indexerArgument = elementReference.Indices[0];
+                        containingType = elementReference.ArrayReference.Type;
+                    }
+                    else if (operationContext.Operation.Kind == OperationKind.None)
+                    {
+                        // The forward support via the "None" operation kind is only available for C#.
+                        if (operationContext.Compilation.Language != LanguageNames.CSharp)
+                        {
+                            return;
+                        }
+
+                        if (operationContext.Operation.Syntax.RawKind != (int)Microsoft.CodeAnalysis.CSharp.SyntaxKind.ElementAccessExpression)
```

Comment:
We should now use `Microsoft.CodeAnalysis.CSharp.SyntaxKind` in shared layer. This will cause C# assemblies to be loaded from VB only solution. If you need to access C# specific types, you should change this type to abstract and have the C# sub-type have C# specific logic.

Reply:
It shouldn't cause an assembly load, since it's an enum (const values are burned in to the caller).  If we're concerned about the ease of creep I could change it to a literal and a comment; which feels better to me than over-engineering a hierarchy here.